broadinstitute/single_cell_portal_core
# Adding Differential Expression study file (SCP-5004)

#### BACKGROUND & CHANGES
This update adds a new `StudyFile` type of "Differential Expression" to allow users to upload their own custom differential expression results.  In addition, an embedded `DifferentialExpressionFileInfo` document has been added to keep track of high-level information about the file, such as the source annotation, computational method type, as well as required associations/validations.  

The update additionally adds the `pairwise_comparisons` field to `DifferentialExpressionResult` entries.  This is a hash of possible combinations of labels detected when the (yet-to-be-implemented) ingest process completes for the user-uploaded file.  This will be used to generate UX elements & file URLs similar to how `observed_values` is used (which denotes "one-vs-rest" comparisons).

Lastly, it also consolidates some common annotation-specific methods into the `Annotatable` concern to reduce code duplication, and fixes a logic bug in validating that source annotations are present.

#### MANUAL TESTING
**Note: if you tested this previously and ran the associated migration, please run this first before continuing:**
```
migration = DataMigration.find_by(version: '20230329181342')
migration.destroy if migration.present?
```
Since there is no upload UI or ingest process for handling these files yet, we can only show that they are valid as `StudyFile` objects:
1. Pull branch, initialize environment, and enter a Rails console session
2. Load the `Human milk - differential expression` study:
```
study = Study.find_by(name: 'Human milk - differential expression')
```
3. Simulate a new file upload with the following block (using a dummy temp file for the upload):
```
FileUtils.cp('test/test_data/table_1.xlsx', 'tmp/table_1.xlsx')
upload = File.open('tmp/table_1.xlsx')
study_file = study.study_files.build(file_type: 'Differential Expression', upload:)
cluster_group = study.default_cluster
annotation_name = 'cell_type__ontology_label'
annotation_scope = 'study'
diff_info = study_file.build_differential_expression_file_info(annotation_name:, annotation_scope:, cluster_group:)
```
4. Save and show the file is valid:
```
study_file.valid?
=> true
study_file.save
=> true
```
5. Inspect the attributes to confirm they are the same (IDs will be different) and then destroy the file to clean up:
```
study_file.attributes
=> 
{"_id"=>BSON::ObjectId('642c4ac894ec8f78cb5401cd'),    
 "parse_status"=>"unparsed",                           
 "x_axis_label"=>"",
 "y_axis_label"=>"",
 "z_axis_label"=>"",
 "spatial_cluster_associations"=>[],
 "remote_location"=>"",
 "options"=>{},
 "file_type"=>"Differential Expression",
 "upload_content_type"=>"application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
 "upload_file_size"=>41692,
 "human_data"=>false,
 "use_metadata_convention"=>false,
 "is_spatial"=>false,
 "queued_for_deletion"=>false,
 "study_id"=>BSON::ObjectId('62718fbb94ec8f51542049d6'),
 "name"=>"table_1.xlsx",
 "data_dir"=>"22e840a57b04cc28a9c0a2b451ea4e6ce14c5d11846c2f2e95fd7033df1169b9",
 "upload_file_name"=>"table_1.xlsx",
 "updated_at"=>2023-04-04 16:06:14.932949 UTC,
 "created_at"=>2023-04-04 16:06:14.932949 UTC,
 "version"=>1,
 "differential_expression_file_info"=>
  {"_id"=>BSON::ObjectId('642c4aca94ec8f78cb5401ce'),
   "computational_method"=>"wilcoxon",
   "annotation_name"=>"cell_type__ontology_label",
   "annotation_scope"=>"study",
   "cluster_group_id"=>BSON::ObjectId('627190374fe67c088c82a84a')}}

study_file.destroy
=> true

```
